### PR TITLE
feat: add support for FlutterAndroidDriver

### PR DESCRIPTION
**Pull request by sudharsan-selvaraj**
## Change list

1. Introduces FlutterAndroidDriver class implementing specific Flutter commands(partial).  
- [X] waitForVisible 
- [X] waitForInvisible
- [X] scrollTillVisible

3. Includes new tests and pipeline updates for running Flutter tests.
4. Note: Changes for IosDriver along with remaining commands will follow in a separate PR.

 
## Types of changes

What types of changes are you proposing/introducing to Java client?
_Put an `x` in the boxes that apply_

- [ ] No changes in production code.
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

Please provide more details about changes if it is necessary. If there are new features you can provide code samples which show the way they
work and possible use cases. Also you can create [gists](https://gist.github.com) with pasted java code samples or put them here using markdown. 
About markdown please read [Mastering markdown](https://guides.github.com/features/mastering-markdown/) and [Writing on GitHub](https://help.github.com/categories/writing-on-github/)

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -89,7 +92,8 @@ jobs:
       if: matrix.e2e-tests == 'android'
       uses: reactivecircus/android-emulator-runner@v2
       with:
-        script: ./gradlew e2eAndroidTest -PisCI -Pselenium.version=$latest_snapshot
+        script: ./gradlew e2eAndroidTest -PisCI -Pselenium.version=$latest_snapshot  &&
```
Comment (mykola-mokhnach) [inline .github/workflows/gradle.yml:95]: I would prefer to separate flutter tests into its own job and add it to the matrix above

File: src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                .withIPAddress("127.0.0.1")
+                .usingPort(PORT)
+                .build();
+        service.start();
+    }
+
+    @BeforeEach
+    public void startSession() throws MalformedURLException {
+        if (IS_ANDROID) {
+            // TODO: update it with FlutterDriverOptions once implemented
```
Comment (mykola-mokhnach) [inline src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java:50]: Are options going to be added in a separate PR?

File: src/main/java/io/appium/java_client/flutter/FlutterDriver.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import io.appium.java_client.flutter.commands.WaitParameter;
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+/**
+ * Interface representing a Flutter driver that extends {@link RemoteWebDriver}.
+ * Provides convenience methods for executing Flutter-specific commands using JavaScriptExecutor.
+ */
+public interface FlutterDriver {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/flutter/FlutterDriver.java:14]: Please group interfaces by their purpose. For example, in this case we may add following interfaces:
- SupportsWaitingForElemements
- SupportsScrolling
- CanExecuteScripts
- HasWebDriver

File: src/main/java/io/appium/java_client/flutter/commands/ScrollParameter.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        this.scrollTo = scrollTo;
+        this.scrollDirection = scrollDirection;
+    }
+
+    @Override
+    public Map<String, Object> toJson() {
+        Map<String, Object> params = new HashMap<>();
+
+        params.put("finder", parseFlutterLocator(scrollTo));
+        params.put("scrollView", scrollView);
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/flutter/commands/ScrollParameter.java:58]: I assume here we must verify if these arguments are not nulls or not and throw errors or skip assigning them

File: src/main/java/io/appium/java_client/flutter/commands/WaitParameter.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+public class WaitParameter extends FlutterCommandParameter {
+    private WebElement element;
+    private AppiumBy.FlutterBy locator;
+    private Duration timeout;
+
+    @Override
+    public Map<String, Object> toJson() {
+        Map<String, Object> args = new HashMap<>();
+        args.put("element", element);
+        if (locator != null) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/flutter/commands/WaitParameter.java:26]: you may use Optional.ofNullable here as well

File: src/main/java/io/appium/java_client/flutter/FlutterDriver.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        String commandName = String.format("flutter: %s", scriptName);
+        return ((JavascriptExecutor) this.getDriver()).executeScript(commandName, parameter.toJson());
+    }
+
+    /**
+     * Retrieves the underlying RemoteWebDriver instance associated with this FlutterDriver.
+     *
+     * @return The RemoteWebDriver instance used by this FlutterDriver.
+     */
+    RemoteWebDriver getDriver();
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/flutter/FlutterDriver.java:61]: I am not sure this method is needed. We could still cast `this` in method calls to whatever we need and it's ok if it throws a runtime exception if the interface is inherited by something that we don't support

File: build.gradle [hunk 7]
```diff
@@ -290,5 +290,27 @@ testing {
...
+        e2eFlutterTest(JvmTestSuite) {
+            sources {
+                java {
+                    srcDirs = ['src/e2eFlutterTest/java']
+                }
+            }
+            dependencies {
+                implementation project()
+                implementation(sourceSets.test.output)
+                implementation('io.github.bonigarcia:webdrivermanager:5.9.1') {
```
Comment (valfirst) [inline build.gradle:303]: it seems this dependency is not used

File: src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            .ofNullable(System.getProperty("platform"))
+            .orElse("android")
+            .equalsIgnoreCase("android");
+    public static final String APP_ID = IS_ANDROID
+            ? "com.example.appium_testing_app" : "com.example.appiumTestingApp";
+    protected static final int PORT = 4723;
+
+    private static AppiumDriverLocalService service;
+    protected static FlutterAndroidDriver driver;
+    protected By loginButton = AppiumBy.flutterText("Login");
```
Comment (valfirst) [inline src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java:33]: it can be static final

File: src/e2eFlutterTest/java/io/appium/java_client/android/CommandTest.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import io.appium.java_client.flutter.commands.ScrollParameter;
+import io.appium.java_client.flutter.commands.WaitParameter;
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.WebElement;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class CommandTest extends BaseFlutterTest {
```
Comment (valfirst) [inline src/e2eFlutterTest/java/io/appium/java_client/android/CommandTest.java:13]: JUnit 5 tests can be package-private

File: src/main/java/io/appium/java_client/AppiumBy.java [hunk 10]
```diff
@@ -206,7 +206,7 @@ public static By iOSNsPredicateString(final String iOSNsPredicateString) {
      * @param selector is the value defined to the key attribute of the flutter element
      * @return an instance of {@link AppiumBy.ByFlutterKey}
      */
-    public static By flutterKey(final String selector) {
+    public static FlutterBy flutterKey(final String selector) {
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/AppiumBy.java:209]: why are these changes needed?

File: src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import java.net.MalformedURLException;
+import java.util.Optional;
+
+public class BaseFlutterTest {
+
+    protected static final boolean IS_ANDROID = Optional
+            .ofNullable(System.getProperty("platform"))
+            .orElse("android")
+            .equalsIgnoreCase("android");
+    public static final String APP_ID = IS_ANDROID
```
Comment (valfirst) [inline src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java:27]: it can be `private`

File: src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.InvalidArgumentException;
+import org.openqa.selenium.WebElement;
+
+import java.net.MalformedURLException;
+import java.util.Optional;
+
+public class BaseFlutterTest {
+
+    protected static final boolean IS_ANDROID = Optional
```
Comment (valfirst) [inline src/e2eFlutterTest/java/io/appium/java_client/android/BaseFlutterTest.java:23]: it can be `private`

File: src/e2eFlutterTest/java/io/appium/java_client/android/CommandTest.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+    AppiumBy.FlutterBy messageFieldLocator = AppiumBy.flutterKey("message_field");
+    AppiumBy.FlutterBy toggleButtonLocator = AppiumBy.flutterKey("toggle_button");
```
Comment (valfirst) [inline src/e2eFlutterTest/java/io/appium/java_client/android/CommandTest.java:16]: these locators can be `private static final`

File: src/main/java/io/appium/java_client/flutter/commands/ScrollParameter.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+@Getter
+@Setter
+public class ScrollParameter extends FlutterCommandParameter {
+    AppiumBy.FlutterBy scrollTo;
+    WebElement scrollView;
+    ScrollDirection scrollDirection;
+    Integer delta;
+    Integer maxScrolls;
+    Integer settleBetweenScrollsTimeout;
+    Duration dragDuration;
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/flutter/commands/ScrollParameter.java:26]: all these fields should be private

**Review by valfirst — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 7 above)
Comment (sudharsan-selvaraj) [reply build.gradle:303]: Nice catch!

File: .github/workflows/gradle.yml (hunk 1 above)
Comment (sudharsan-selvaraj) [reply .github/workflows/gradle.yml:95]: Got it. I'll update the pipline with new jobs for flutter tests.

File: src/main/java/io/appium/java_client/AppiumBy.java (hunk 10 above)
Comment (sudharsan-selvaraj) [reply src/main/java/io/appium/java_client/AppiumBy.java:209]: Currenlty all flutter script can work with locatar supported by flutter driver. This change is specifically introduced the simplify the API consumption and to prevent the manual type casting when calling the commands.

```java
driver.waitForInVisible(new WaitParameter().setLocator(AppiumBy.flutterKey("message_field")));
```

Instead of 

```java
driver.waitForInVisible(new WaitParameter().setLocator((AppiumBy.FlutterBy)AppiumBy.flutterKey("message_field")));
```

**Review by mykola-mokhnach — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Review by SrinivasanTarget — APPROVED** [review]
